### PR TITLE
Improve pruning sparsity levels selection for the pruning modifier

### PR DESCRIPTION
**Pull request by markurtz**
- base off of a combination of weight magnitude and perf sensitivities
- fixes so the algorithm will run quickly and correctly
- base case should match a weight magnitude selection
- change recovery to compare against baseline of weight magnitude without any restrictions

### CODE REVIEW

File: src/sparsify/blueprints/utils/projects_optimizations_pruning.py [hunk 1]
```diff
@@ -51,290 +49,228 @@
...
-        rescaled = val - self._avg_mins
-        if self._avg_ranges:
-            rescaled = rescaled / self._avg_ranges
+        # normalize the value such that it will fall in the range [0, max]
+        # by subtracting the min
+        rescaled = val - self.min_val if self.min_val else val
+
+        # normalize the value such that it will fall in the range [0, 1.0]
+        # by dividing by the max range
+        max_range = self.max_val - self.min_val
```
Comment (natuan) [inline src/sparsify/blueprints/utils/projects_optimizations_pruning.py:74]: Will "add_rescale_series" always be called be for this method? Can self.max_val be None here? (There's a None check for min above, but not for max)

Comment (markurtz) [reply src/sparsify/blueprints/utils/projects_optimizations_pruning.py]: fixed

**Review by mgoin — DISMISSED** [review]
nice, LGTM

File: src/sparsify/blueprints/utils/projects_optimizations_pruning.py [hunk 2]
```diff
@@ -51,290 +49,232 @@
...
-    ) -> Union[None, float]:
-        """
-        :param sparsity: the sparsity to get the gain value for
-        :param smooth: True to pull from the measurements_smoothed,
-            False otherwise
-        :return: the ratio of the predicted value at the given sparsity
-            as compared with the baseline value
-        """
-        if not self._measurements:
+    def estimated_gain(self, sparsity: Union[None, float]) -> Union[None, float]:
```
Comment (natuan) [inline src/sparsify/blueprints/utils/projects_optimizations_pruning.py:159]: Docstring for this func (from sparse_gain?)

File: src/sparsify/blueprints/utils/projects_optimizations_pruning.py [hunk 3]
```diff
@@ -51,290 +49,232 @@
...
-        """
-        :param sparsity: the sparsity to get the sensitivity value for
-        :param smooth: True to pull from the measurements_smoothed,
-            False otherwise
-        :return: the sensitivity comparison (difference) of the measurement
-            at the given sparsity compared with the baseline
-        """
-        sparse = self.sparse(sparsity, smooth)
-        baseline = self.baseline
+    def estimated_sensitivity(self, sparsity: Union[None, float]) -> Union[None, float]:
```
Comment (natuan) [inline src/sparsify/blueprints/utils/projects_optimizations_pruning.py:173]: docstring (from sparse_sensitivity?)

**Review by natuan — COMMENTED** [review]
NIT: some functions I think are helpful to add docstring

**Comment by markurtz** [comment]
@natuan everything marked with docstrings have been moved to private classes that are not exported from the file. Per our standards, are the docstrings required for understanding the function of the private classes/methods?

**Comment by natuan** [comment]
> @natuan everything marked with docstrings have been moved to private classes that are not exported from the file. Per our standards, are the docstrings required for understanding the function of the private classes/methods?

I noticed those classes are private. At the same time I'm not sure I agree all functions/methods then shouldn't need docstrings or explanation (noticed I mentioned just 3-4 functions that I found their original docs helpful for me to understand). Private classes could be complicated.

**Comment by markurtz** [comment]
> > @natuan everything marked with docstrings have been moved to private classes that are not exported from the file. Per our standards, are the docstrings required for understanding the function of the private classes/methods?
> 
> I noticed those classes are private. At the same time I'm not sure I agree all functions/methods then shouldn't need docstrings or explanation (noticed I mentioned just 3-4 functions that I found their original docs helpful for me to understand). Private classes could be complicated.

Yes, was making sure the move from public to private classes was noticed for them and the ask for docs were for clarification and not on the assumption of public. Our standards are to default to no docstrings for private defs/classes unless requested for understanding/reading. Will update the docs for those statements

**Comment by natuan** [comment]
> > > @natuan everything marked with docstrings have been moved to private classes that are not exported from the file. Per our standards, are the docstrings required for understanding the function of the private classes/methods?
> > 
> > 
> > I noticed those classes are private. At the same time I'm not sure I agree all functions/methods then shouldn't need docstrings or explanation (noticed I mentioned just 3-4 functions that I found their original docs helpful for me to understand). Private classes could be complicated.
> 
> Yes, was making sure the move from public to private classes was noticed for them and the ask for docs were for clarification and not on the assumption of public. Our standards are to default to no docstrings for private defs/classes unless requested for understanding/reading since those docs won't get compiled through sphinx for the docs website. Will update the docs for those statements

Sounds good to me. Thanks!

**Review by natuan — APPROVED** [review]

**Review by kevinaer — APPROVED** [review]
LGTM. Small comment I've been meaning to bring up but I noticed pytorch is using Optional[type] instead of Union[type, None]. What are thoughts on using Optional instead?

**Comment by markurtz** [comment]
> LGTM. Small comment I've been meaning to bring up but I noticed pytorch is using Optional[type] instead of Union[type, None]. What are thoughts on using Optional instead?

They both equate to the same thing for typing systems and I don't have a preference one way or the other. Happy to talk over what everyone prefers as a standard to set / if we should set a standard for this

**Pull request merged**